### PR TITLE
Add docker generation script

### PR DESCRIPTION
**Pull request by keithc-ca**
Add Dockerfile generation script:
* add support for Ubuntu 18.04 and 20.04
* use api.adoptopenjdk.net/v3 to fetch boot JDKs
* install gcc version 7.5 on CentOS from adoptopenjdk
* install git version 2.5.3 from source on CentOS
* install make 4.2 for ppc64le on CentOS
* install nasm version 2.13.03 from source on CentOS for x86

Remove Dockerfiles obsoleted by mkdocker.sh and relocate ARM Dockerfiles.

Update references to obsolete Dockerfiles.

Simplify ansible Dockerfile.

Fixes: #1575
Fixes: #5781
Fixes: #6767
Fixes: #9057

### CODE REVIEW

**Comment by keithc-ca** [comment]
Usage as described via the `--help` option
```
$ bash mkdocker.sh --help
Usage: mkdocker.sh [OPTION]...
Generate a Dockerfile for building OpenJ9

Options:
  --help|-h      print this help, then exit
  --arch=...     specify the processor architecture (default: host architecture)
  --build        build the docker image (overrides '--print')
  --cuda         include CUDA header files
  --dist=...     specify the Linux distribution (e.g. centos, ubuntu)
  --print        write the Dockerfile to stdout (default; overrides '--build')
  --tag=...      specify a name for the docker image (may be repeated, default: none)
  --user=...     specify the user name (default: 'jenkins')
  --version=...  specify the distribution version (e.g. 6.9, 16.04)

Supported build patterns on this host (x86_64):
  bash mkdocker.sh --tag=openj9/cent69 --dist=centos --version=6.9 --build
  bash mkdocker.sh --tag=openj9/cent7  --dist=centos --version=7   --build
  bash mkdocker.sh --tag=openj9/ub16   --dist=ubuntu --version=16  --build
  bash mkdocker.sh --tag=openj9/ub18   --dist=ubuntu --version=18  --build
  bash mkdocker.sh --tag=openj9/ub20   --dist=ubuntu --version=20  --build
```

File: buildenv/ansible/roles/openj9-build-environment/tasks/main.yml [hunk 1]
```diff
@@ -22,20 +22,22 @@
...
-    mode: 0644
+    dest: "{{ build_dir }}/mkdocker.sh"
+    url: "https://raw.githubusercontent.com/eclipse/openj9/master/buildenv/docker/mkdocker.sh"
+    mode: 0755
     validate_certs: no
   tags:
     - buildenv
 - name: Build the docker image for the building environment
+  # docker/{{ version }}/{{ platform }}/ubuntu16
+  # bash mkdocker.sh --tag=openj9-{{ version }}-build --dist=ubuntu --version=16.04 --build
```
Comment (pshipton) [inline buildenv/ansible/roles/openj9-build-environment/tasks/main.yml:40]: Is this supposed to be a comment?

File: buildenv/jenkins/docker-slaves/jenkinsfile-build-container.groovy [hunk 2]
```diff
@@ -63,17 +63,24 @@ timeout(time: 5, unit: 'HOURS') {
                 }
                 stage("Build") {
                     if(params.ghprbPullId) {
-                        TAGS = "-t ${REPOSITORY}:PR${ghprbPullId}"
+                        TAGS = "--tag=${REPOSITORY}:PR${ghprbPullId}"
                     } else {
-                        TAGS = "-t ${REPOSITORY}:${BUILD_NUMBER} -t ${REPOSITORY}:latest"
+                        /* FIXME mkdocker.sh doesn't handle multiple --tag options */
```
Comment (pshipton) [inline buildenv/jenkins/docker-slaves/jenkinsfile-build-container.groovy:68]: There should be an Issue for this.

File: doc/build-instructions/Build_Instructions_V8.md [hunk 3]
```diff
@@ -159,10 +125,8 @@ bash get_source.sh
 :penguin:
 When you have all the source files that you need, run the configure script, which detects how to build in the current build environment.
 ```
-bash configure --with-freemarker-jar=/root/freemarker.jar
+bash configure --with-cmake
```
Comment (pshipton) [inline doc/build-instructions/Build_Instructions_V8.md:128]: We're not ready for this yet.

File: buildenv/ansible/roles/openj9-build-environment/tasks/main.yml [hunk 4]
```diff
@@ -1,5 +1,5 @@
 ###############################################################################
-# Copyright (c) 2018, 2018 Pavel Samolysov
+# Copyright (c) 2018, 2020 Pavel Samolysov
```
Comment (pshipton) [inline buildenv/ansible/roles/openj9-build-environment/tasks/main.yml:2]: This isn't an IBM copyright, seems an IBM copyright needs to be added for these changes.

**Comment by pshipton** [comment]
> centos:7 is only available for ppc64le and x86_64

Note support is required for centos:7 on s390x, considering this is the build platform. I've gone ahead and created https://github.com/eclipse/openj9/issues/9372

File: buildenv/jenkins/docker-slaves/jenkinsfile-build-container.groovy (hunk 2 above)
Comment (keithc-ca) [reply buildenv/jenkins/docker-slaves/jenkinsfile-build-container.groovy]: The script already supports multiple `--tag` options: the comment was just stale and has been removed.

**Comment by pshipton** [comment]
Should probably add `[ci skip]` to the commit comment for next time.

File: buildenv/ansible/roles/openj9-build-environment/tasks/main.yml (hunk 1 above)
Comment (keithc-ca) [reply buildenv/ansible/roles/openj9-build-environment/tasks/main.yml:40]: No, it is a reminder that I need to learn enough about ansible to run that `bash` command.

File: doc/build-instructions/Build_Instructions_V8.md (hunk 3 above)
Comment (keithc-ca) [reply doc/build-instructions/Build_Instructions_V8.md:128]: I thought we were close enough (at least for jdk8 on Linux) that it made sense to update this, but I will revert to non-cmake for now.

**Comment by pshipton** [comment]
@knn-k fyi

**Comment by pshipton** [comment]
@AdamBrousseau fyi

File: buildenv/ansible/roles/openj9-build-environment/tasks/main.yml (hunk 1 above)
Comment (pshipton) [reply buildenv/ansible/roles/openj9-build-environment/tasks/main.yml:40]: To merge in advance of fixing this problem, we should have an issue to address it later.

**Comment by pshipton** [comment]
I note only the Build_Instructions for V8 are updated. Are you intending to update the others, or expecting I'll update them afterwards (since I volunteered)?

File: buildenv/ansible/roles/openj9-build-environment/tasks/main.yml (hunk 1 above)
Comment (keithc-ca) [reply buildenv/ansible/roles/openj9-build-environment/tasks/main.yml]: It turned out to be easier to just fix it.

File: buildenv/ansible/roles/openj9-build-environment/tasks/main.yml (hunk 4 above)
Comment (keithc-ca) [reply buildenv/ansible/roles/openj9-build-environment/tasks/main.yml]: Fixed.

File: buildenv/ansible/roles/openj9-build-environment/tasks/main.yml [hunk 5]
```diff
@@ -19,40 +19,36 @@
 #
 # SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
 ###############################################################################
+# (c) Copyright IBM Corp. 2020, 2020 All Rights Reserved
```
Comment (pshipton) [inline buildenv/ansible/roles/openj9-build-environment/tasks/main.yml:22]: Pls use the standard ` Copyright (c) 2020, 2020 IBM Corp. and others` copyright, used by most of the OpenJ9 files.

File: doc/build-instructions/Build_Instructions_V8.md (hunk 3 above)
Comment (keithc-ca) [reply doc/build-instructions/Build_Instructions_V8.md]: Reverted.

**Comment by keithc-ca** [comment]
> I note only the Build_Instructions for V8 are updated. Are you intending to update the others, or expecting I'll update them afterwards (since I volunteered)?

My plan is to update them later (with your help).

File: buildenv/ansible/roles/openj9-build-environment/tasks/main.yml (hunk 5 above)
Comment (keithc-ca) [reply buildenv/ansible/roles/openj9-build-environment/tasks/main.yml]: I copied that text from some other file (in jdk8?). Updated as requested.

Comment (pshipton) [reply buildenv/ansible/roles/openj9-build-environment/tasks/main.yml]: Yes, the ibmruntimes repos use different copyrights.

File: buildenv/docker/mkdocker.sh [hunk 6]
```diff
@@ -0,0 +1,672 @@
...
+
+  install_compilers
+
+  install_cmake
+  install_python
+
+  adjust_ldconfig
+  configure_ssh
+
+  install_bootjdks
```
Comment (fjeremic) [inline buildenv/docker/mkdocker.sh:656]: This seems to download quite a few JDKs. Is the intention that the resulting Dockerfile will be able to build all versions of the JDK? If so does that mean the user is expected to explicitly initialize `JAVA_HOME` to point to the correct JDK before kicking off the build process?

Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh:656]: Yes, boot JDKs will be available for building any (current) version. Users are expected to use the `--with-boot-jdk=` configure option, e.g.:
```
bash configure --with-boot-jdk=/usr/lib/jvm/adoptojdk-java-80 ...
```
or
```
bash configure --with-boot-jdk=/usr/lib/jvm/adoptojdk-java-11 ...
```
Setting `JAVA_HOME` is not necessary.

**Comment by keithc-ca** [comment]
> Should probably add `[ci skip]` to the commit comment for next time.

I usually forget to do that. Done now.

File: buildenv/docker/mkdocker.sh (hunk 6 above)
Comment (pshipton) [reply buildenv/docker/mkdocker.sh:656]: I use the `--with-boot-jdk=` configure option, but I see we don't consistently suggest that in the build docs, it's only suggested for Windows.

Comment (pshipton) [reply buildenv/docker/mkdocker.sh:656]: Created https://github.com/eclipse/openj9/issues/9375 to deal with this problem.

**Comment by pshipton** [comment]
I'll wait until tomorrow before merging in case there are other comments.

**Comment by pshipton** [comment]
I think ant is missing from centos installs?

**Comment by keithc-ca** [comment]
> I think ant is missing from centos installs?

Alternatively, it doesn't belong in ubuntu images, and should be the responsibility of [buildenv/docker/test/Dockerfile](https://github.com/eclipse/openj9/blob/master/buildenv/docker/test/Dockerfile) to make `ant` available?

Let me know your preference.

**Comment by pshipton** [comment]
Since the script is replacing the jenkins/docker-slaves/ Dockerfiles's, and these are setup to do build and test, ant needs to be installed. Plus anything else required for test that I missed.

**Comment by keithc-ca** [comment]
Updated to install `ant` and `ant-contrib` on CentOS.

**Comment by pshipton** [comment]
I already had a CentOS image so I ran the commands manually in it.
```
mv ant-contrib/ant-contrib-1.0b3.jar /opt/ant/lib
mv: cannot move 'ant-contrib/ant-contrib-1.0b3.jar' to '/opt/ant/lib': No such file or directory
```

**Comment by pshipton** [comment]
Never mind, I see where I went wrong.

**Comment by pshipton** [comment]
I ran the image as suggested in the guide `docker run -v <host_directory>:/root/hostdir -it openj9`. /root/hostdir isn't accessible to the default user `root`. 
`rwx------. 53 jenkins 1001  16384 Apr 28 02:23 hostdir`

**Comment by knn-k** [comment]
Re: renaming the path of Dockerfile for AArch64 cross compilation.

I used the directory name `aarch64_CC/arm-linux-aarch64` following that of existing 32-bit ARM `armhf_CC/arm-linux-gnueabihf`, where `CC` stands for cross compilation.  Now I think `arm-linux-aarch64` is not an appropriate name.
How about changing it from `aarch64_CC/arm-linux-aarch64` to `aarch64_linux_CC`, one fewer depth?
The path for 32-bit ARM can be `armhf_linux_CC` if changing it in the same way.

**Comment by knn-k** [comment]
In addition, I have never tried building JDK 14 using the Docker image for AArch64 cross compilation.

File: buildenv/docker/mkdocker.sh [hunk 7]
```diff
@@ -0,0 +1,684 @@
...
+  echo "RUN cd /tmp \\"
+  echo " && $wget_O make.tar.gz https://github.com/mirror/make/archive/$make_version.tar.gz \\"
+  echo " && tar -xzf make.tar.gz \\"
+  echo " && cd make-$make_version \\"
+  echo " && ACLOCAL_PATH=/usr/share/aclocal autoreconf -i \\"
+  echo " && ./configure \\"
+  echo " && make update \\"
+  echo " && make \\"
+  echo " && make install \\"
+  echo " && ln -s make /usr/local/bin/gmake \\"
```
Comment (pshipton) [inline buildenv/docker/mkdocker.sh:365]: I think something is wrong here. The image runs make 3.82 and there is no `/usr/local/bin/gmake` file.

**Comment by AdamBrousseau** [comment]
Jenkins build docker x86 centos6.9

**Comment by keithc-ca** [comment]
> Re: renaming the path of Dockerfile for AArch64 cross compilation.
> 
> I used the directory name `aarch64_CC/arm-linux-aarch64` following that of existing 32-bit ARM `armhf_CC/arm-linux-gnueabihf`, where `CC` stands for cross compilation. Now I think `arm-linux-aarch64` is not an appropriate name.
> How about changing it from `aarch64_CC/arm-linux-aarch64` to `aarch64_linux_CC`, one fewer depth?
> The path for 32-bit ARM can be `armhf_linux_CC` if changing it in the same way.

I only removed the `jdk9` and `jdk11` segments of the pathnames of those files. I suggest any further name changes should be done in a separate pull request. The only 'jdk*' folder left is the `jdk8` one that contains JIT Dockerfiles that may also be obsolete.

File: buildenv/docker/mkdocker.sh (hunk 7 above)
Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh:365]: This is only used on ppc64le where the version of make available via apt appears to hang during a build.
```
keithc@up16keithc1:~$ docker run --rm -it openj9/cent7
[root@b645b703025b /]# which make
/usr/local/bin/make
[root@b645b703025b /]# make -v
GNU Make 4.2
Built for powerpc64le-unknown-linux-gnu
Copyright (C) 1988-2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

File: buildenv/jenkins/docker-slaves/jenkinsfile-build-container.groovy [hunk 8]
```diff
@@ -63,17 +63,23 @@ timeout(time: 5, unit: 'HOURS') {
...
+                        sh "touch known_hosts"
+                        if(env.KNOWN_HOSTS) {
+                            sh "ssh-keyscan ${KNOWN_HOSTS} >> known_hosts"
+                        }
+                        if(${ARCH} == "ppc64le") {
+                            sh "bash mkdocker.sh --build --dist=centos --version=7 ${TAGS}"
+                        } else if(${ARCH} == "s390x") {
+                            sh "bash mkdocker.sh --build --dist=ubuntu --version=16.04 ${TAGS}"
+                        } else if(${ARCH} == "x86") {
+                            sh "bash mkdocker.sh --build --dist=centos --version=6.9 ${TAGS}"
```
[the remark this answers is not in the document]
Comment (keithc-ca) [reply buildenv/jenkins/docker-slaves/jenkinsfile-build-container.groovy:81]: How would an `x86_64` host, say, build and image for `ppc64le`? I don't see how it makes sense to expect to be able to do cross-platform _builds_. On the other hand, if we only wanted to _print_ a `Dockerfile` for a different OS, that's fine, but I don't think that was what this was trying to do.

I think test-only images make sense to verify that none of the _build_ requirements become necessary to _use_ a VM.

File: buildenv/docker/mkdocker.sh [hunk 9]
```diff
@@ -0,0 +1,684 @@
...
+if [ ${#tags[@]} -eq 0 ] ; then
+  echo "#   docker run -it <image>"
+else
+for tag in ${tags[@]} ; do
+  echo "#   docker run -it $tag"
+done
+fi
+  echo ""
+if [ $cuda != no ] ; then
+  echo "FROM nvidia/cuda:${cuda}-devel-ubuntu16.04 AS cuda-dev"
```
[the remark this answers is not in the document]
Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh:219]: No, we're only copying a few header files from the CUDA image; they're all platform-independent.

File: buildenv/jenkins/docker-slaves/jenkinsfile-build-container.groovy (hunk 8 above)
[the remark this answers is not in the document]
Comment (keithc-ca) [reply buildenv/jenkins/docker-slaves/jenkinsfile-build-container.groovy:81]: Perhaps I was too hasty in my reply above. Do we need to build an image for any of the existing `ARCH/OS` combinations? If so, I can update this file so it matches the old behavior.
```
ppc64le/centos7
ppc64le/ubuntu16
ppc64le/ubuntu18
s390x/ubuntu16
s390x/ubuntu18
x86/centos6.9
x86/ubuntu16
x86/ubuntu18
```

**Comment by knn-k** [comment]
@keithc-ca 
OK, I will rename Dockerfiles for cross compilation (arm, aarch64) later separately.

File: buildenv/jenkins/docker-slaves/jenkinsfile-build-container.groovy (hunk 8 above)
[the remark this answers is not in the document]
Comment (keithc-ca) [reply buildenv/jenkins/docker-slaves/jenkinsfile-build-container.groovy]: I've updated the script to support the expanded set of combinations handled by `mkdocker.sh`.

**Comment by keithc-ca** [comment]
Jenkins build docker x86 centos6.9

File: buildenv/docker/mkdocker.sh (hunk 7 above)
Comment (pshipton) [reply buildenv/docker/mkdocker.sh:365]: The tests require make 4.1 or later. The cent7/x86 image contains make 3.82

Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh:365]: I'll fix that, but where does it say the tests need 4.1 (besides perhaps ppc64le)?

Comment (pshipton) [reply buildenv/docker/mkdocker.sh:365]: The test requirements are here https://github.com/eclipse/openj9/blob/master/test/docs/Prerequisites.md

Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh:381]: Updated to install make 4.1 on CentOS for x86_64.

**Comment by pshipton** [comment]
@AdamBrousseau do you have any further concerns?

**Comment by keithc-ca** [comment]
Hold off for now: update make to 4.1 on centos:6.9 broke centos:7 on x86.

File: buildenv/jenkins/docker-slaves/jenkinsfile-build-container.groovy (hunk 8 above)
[the remark this answers is not in the document]
Comment (AdamBrousseau) [reply buildenv/jenkins/docker-slaves/jenkinsfile-build-container.groovy]: Great thanks. Looks like there's no notion of test-only at this point. Which I guess is more user/developer friendly. We are not using the existing images at the moment (anymore) but I would like to re-enable them. We may only do 1 for each arch to start in order to limit our risk for stability issues.

**Comment by keithc-ca** [comment]
All the CentOS variants are working again.

File: buildenv/docker/mkdocker.sh [hunk 10]
```diff
@@ -0,0 +1,697 @@
...
+  echo " && chmod 700 /home/$user/.ssh"
+  echo "COPY authorized_keys known_hosts /home/$user/.ssh/"
+  echo "RUN chown -R $user:$user /home/$user \\"
+  echo " && chmod 600 /home/$user/.ssh/authorized_keys /home/$user/.ssh/known_hosts"
+}
+
+install_freemarker() {
+  echo ""
+  local freemarker_version=2.3.8
+  echo "# Download and extract freemarker.jar to /root/freemarker.jar."
```
[the remark this answers is not in the document]
Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh:544]: All the Dockerfiles under buildenv/docker put it in /root and [ansible](https://github.com/eclipse/openj9/blob/master/buildenv/ansible/roles/openj9-from-sources/tasks/main.yml#L57) does the same. On the other hand, [defaults.yml](https://github.com/eclipse/openj9/blob/master/buildenv/jenkins/variables/defaults.yml#L146) seems to expect it in /home/jenkins. I suppose we could have it both places, or we can pick one and fix the references to the other location.

File: buildenv/docker/mkdocker.sh [hunk 11]
```diff
@@ -0,0 +1,697 @@
...
+
+create_git_cache() {
+  # The jdk14 remote is fetched first because that repository was subjected to
+  # 'git gc --aggressive --prune=all' before it was first published making it much
+  # smaller than the other jdk repositories. There is a large degree of overlap
+  # among the jdk repositories so a relatively small number of commits must be
+  # fetched from the others.
+  echo ""
+  echo "# Setup a reference repository cache for faster clones in containers."
+  echo "RUN mkdir /home/$user/git_cache \\"
```
[the remark this answers is not in the document]
Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh:632]: This is easily changed once we have agreement on what it should be called.

File: buildenv/docker/mkdocker.sh [hunk 12]
```diff
@@ -0,0 +1,697 @@
...
+  echo "    libXt-devel \\"
+  echo "    libXtst-devel \\"
+  echo "    make \\"
+  echo "    mesa-libGL-devel \\"
+  echo "    mpfr-devel \\"
+  echo "    ntp \\"
+  echo "    numactl-devel \\"
+  echo "    openssh-clients \\"
+  echo "    openssh-server \\"
+  echo "    openssl-devel \\" # required by git, etc.
```
[the remark this answers is not in the document]
Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh:273]: Which packages are you suggesting might be omitted?

The version of openssl use to build is less important than what is bundled.

[the remark this answers is not in the document]
Comment (AdamBrousseau) [reply buildenv/docker/mkdocker.sh:273]: On the existing instead we specifically install 111b.
https://github.com/eclipse/openj9/blob/master/buildenv/jenkins/docker-slaves/x86/centos6.9/Dockerfile#L244
But that may be independent now.

[the remark this answers is not in the document]
Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh:273]: We're at openssl 1.1.1g currently.

A build of jdk11 configured `--with-jitserver` in a centos:6.9/x86 container completed successfully.

[the remark this answers is not in the document]
Comment (pshipton) [reply buildenv/docker/mkdocker.sh:273]: Any Linux system installed openssl satisfies a runtime requirement, not a build requirement. The build downloads 1.1.1g (currently) and builds against that, it doesn't matter what is installed on the system.
i.e. you'd need to run JIT server testing to see any problem, which may include using `test_flags: 'JITAAS'`. I'm not sure if existing tests would identify problems with openssl. Better to ask @mpirvu about the requirements.

[the remark this answers is not in the document]
Comment (pshipton) [reply buildenv/docker/mkdocker.sh:273]: I note https://blog.openj9.org/2020/01/09/free-your-jvm-from-the-jit-with-jitserver-technology/ mentions `If so desired, the network communication between the client JVM and JITServer can be encrypted using OpenSSL 1.0.x or 1.1.x`

[the remark this answers is not in the document]
Comment (mpirvu) [reply buildenv/docker/mkdocker.sh:273]: yes, JITServer can build against OpenSSL 1.1.x, so we should be good.

[the remark this answers is not in the document]
Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh:273]: I don't think we should hold this up for JITServer requirements: there are still a bunch of dockerfiles that serve those purposes:
```
buildenv/docker/jdk8/s390x/ubuntu16/jitserver
buildenv/docker/jdk8/x86_64/rhel7/jitserver
buildenv/docker/jdk8/x86_64/ubuntu16/jitserver
buildenv/docker/jdk8/x86_64/ubuntu18/jitserver
```
I hope we will come to decide that most of them are redundant (or the script can be improved to make them so).

[the remark this answers is not in the document]
Comment (pshipton) [reply buildenv/docker/mkdocker.sh:273]: Agreed, this isn't a blocker for merging. I think the git_cache issue is. The other stuff can be fixed up afterwards if necessary.

File: buildenv/docker/mkdocker.sh (hunk 11 above)
[the remark this answers is not in the document]
Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh:632]: I give in: let's stick with 'openjdk_cache' (at least for the time-being).

**Comment by keithc-ca** [comment]
Updated git cache directory to match current usage (`openjdk_cache`).

**Comment by pshipton** [comment]
Any other concerns can be addressed in another PR.

**Pull request merged**